### PR TITLE
Including subpages feature on pages field

### PR DESCRIPTION
**Pull request by afbora**
## Describe the PR

Allows you to include sub-pages with using query or not

I didn't want to separate options as `subpages` and `querySubpages` as @lukasbestle suggested. This could create a bit of confusion for users and code
I wanted to make a simpler and more comprehensive (with query or not) implement with one option as `subpages`
Including subpages default option is enabled and to disable set as `subpages: false`

**Usage**:

```yaml
fields:
    testWithDefault:
        type: pages
        multiple: true
        
    testWithoutSubpages:
        type: pages
        multiple: true
        subpages: false
        
    testWithQuerySubpages:
        type: pages
        multiple: true
        query: site.find("test").children
```

**Note**:
@lukasbestle I know i can't write a very effective unit test. I think a different method needs to be applied for child elements. Therefore, it may be necessary to review the `testPageChildren()` test 😇 

**Additional idea**: 
In addition, you can select pages from specific levels by adding depth option. If approval is given, I can implement this idea. For ex:

```yaml
fields:
    testWithDepth:
        type: pages
        multiple: true
        depth: 2

```

## Related issues

- Closes https://github.com/getkirby/ideas/issues/217

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Fixed code style issues with CS fixer and `composer fix`
- [x] Added in-code documentation (if neeed)

### CODE REVIEW

File: config/fields/mixins/pagepicker.php [hunk 1]
```diff
@@ -3,23 +3,35 @@
...
+            $model    = $this->model();
+            $site     = $this->kirby()->site();
 
-            if ($query) {
-                $pages = $model->query($query, 'Kirby\Cms\Pages');
-                $self  = null;
+            if (empty($query) === false) {
+                if ($subpages === true) {
+                    if ($parent = $site->find($params['parent'] ?? null)) {
+                        $pages = $parent->children();
```
Comment (lukasbestle) [inline config/fields/mixins/pagepicker.php]: Why the check for the `parent` option here? Shouldn't the query always be executed if it's given? I'd say this is a breaking change otherwise.

File: config/fields/mixins/pagepicker.php [hunk 2]
```diff
@@ -3,23 +3,35 @@
...
+                    } else {
+                        $pages  = $model->query($query, 'Kirby\Cms\Pages');
+                        $parent = $pages->parent();
+                    }
+                } else {
+                    $pages = $model->query($query, 'Kirby\Cms\Pages');
```
Comment (lukasbestle) [inline config/fields/mixins/pagepicker.php]: `$parent` should explicitly be set to `null` here to ensure that the variable is set in all cases of the condition tree.

File: config/fields/mixins/pagepicker.php [hunk 3]
```diff
@@ -3,23 +3,35 @@
...
             } else {
                 if (!$parent = $site->find($params['parent'] ?? null)) {
                     $parent = $site;
                 }
-
+                
                 $pages = $parent->children();
+            }
+            
+            if ($subpages === true) {
```
Comment (lukasbestle) [inline config/fields/mixins/pagepicker.php]: This should then check for `$subpages === true && $parent !== null` to make sure that the following code never runs without a valid parent object.

File: config/fields/mixins/pagepicker.php [hunk 4]
```diff
@@ -3,23 +3,35 @@
...
                 $self  = [
-                    'id'     => $parent->id() == '' ? null : $parent->id(),
-                    'title'  => $parent->title()->value(),
+                    'id'     => (empty($parent->id()) === true || empty($params['parent']) === true) ? null : $parent->id(),
```
Comment (lukasbestle) [inline config/fields/mixins/pagepicker.php]: Why the check for the `$params['parent']` value?

File: config/fields/pages.php [hunk 5]
```diff
@@ -42,6 +42,13 @@
         'size' => function (string $size = 'auto') {
             return $size;
         },
+        
+        /**
+         * Optional includes subpages of pages
```
Comment (lukasbestle) [inline config/fields/pages.php]: "Optionally include subpages of pages"

**Review by lukasbestle — DISMISSED** [review]

File: config/fields/mixins/pagepicker.php (hunk 4 above)
Comment (afbora) [reply config/fields/mixins/pagepicker.php]: FireShot Capture 006 - Sandbox - Mægazine - localhost

This is for hide back button for root.

File: config/fields/mixins/pagepicker.php (hunk 1 above)
Comment (afbora) [reply config/fields/mixins/pagepicker.php]: In the first query/load, `parent` index is `null`. When navigating to the parent and subpages, the `parent` index is the `id` of the clicked page.

File: config/fields/mixins/pagepicker.php [hunk 6]
```diff
@@ -3,26 +3,42 @@
...
-            $site  = $this->kirby()->site();
-
-            if ($query) {
-                $pages = $model->query($query, 'Kirby\Cms\Pages');
-                $self  = null;
+            $query    = $params['query'] ?? null;
+            $subpages = $params['subpages'] ?? true;
+            $model    = $this->model();
+            $site     = $this->kirby()->site();
+            $root     = $site;
```
Comment (afbora) [inline config/fields/mixins/pagepicker.php]: FireShot Capture 006 - Sandbox - Mægazine - localhost

This page is root but showing back button.
Therefore keeping `root` page for show/hide back button.

File: config/fields/mixins/pagepicker.php [hunk 7]
```diff
@@ -3,26 +3,42 @@
...
-
+                
                 $pages = $parent->children();
+            }
+            
+            if ($subpages === true && $parent !== null) {
                 $self  = [
-                    'id'     => $parent->id() == '' ? null : $parent->id(),
-                    'title'  => $parent->title()->value(),
+                    'id'     => (empty($parent->id()) === true || $parent->id() === $root->parent()->id()) ? null : $parent->id(),
```
Comment (afbora) [inline config/fields/mixins/pagepicker.php]: This is for show/hide back button.

**Review by lukasbestle — COMMENTED** [review]
The logic is quite complex, which makes it hard to review if the code works for all edge-cases. I'd say this needs a unit test for each branch in the `if` tree.

Also this is probably something for @bastianallgeier to review as he wrote the original code. It's quite hard for me to follow the logic.

**Comment by bastianallgeier** [comment]
I'm on it. Maybe we can refactor it a bit to be more readable. It has been my spaghetti code for a long time.

**Comment by bastianallgeier** [comment]
I moved the logic for the picker into its own class now. This makes it more future-proof. Especially when we want to work on pagination and search as a next step.
